File: Data Structures/Stacks/Stack.cpp
#include <iostream>
using namespace std;
struct Node{
    int data;
    struct Node* next;
};
class Stack{
    public:
    Node *head, *tail;
    Stack(){
        head = NULL;
        tail = NULL;
    }
    void push(int data){
        Node *temp = new Node();
        temp->data = data;
        if(head == NULL){
            head = temp;
            tail = head;
            return;
        }
        tail->next = temp;
        tail = tail->next;
    }
    void pop(){
        Node *cur, *precur;
        cur = head;
        while(cur != NULL){
            precur = cur;
            cur = cur->next;
        }
        tail = precur;
    }
    void print(){
        Node *temp = head;
        while(temp != NULL){
            cout<<temp->data<<endl;
            temp = temp->next;
        }
    }
};
int main(){
    Stack s;
    s.push(1);
    s.push(2);
    s.push(3);
    s.push(4);
    s.push(5);
    s.print();
    s.pop();
    s.print();
    return 0;
}

File: Data Structures/Stacks/stack.py
class Node:

    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class Stack:

    def __init__(self):
        self.root = None
    
    def push(self, data):
        temp = Node(data)
        if self.root is None:
            self.root = temp
        else:
            current = self.root
            while current.next is not None:
                current = current.next
            current.next = temp
            temp.prev = current

    def pop(self):
        if self.root is None:
            try:
                raise Exception('Stack Underflow')
            except Exception as e:
                print(e)
        if self.root.next is None:
            res = self.root.data
            self.root = None
            return res
        ptr = self.root.next
        preptr = self.root
        while ptr.next is not None:
            ptr = ptr.next
            preptr = preptr.next
        preptr.next = None
        return ptr.data
    
    def print(self):
        current = self.root
        while current is not None:
            print(current.data, end=' ')
            current = current.next
        print()
    
if __name__ == "__main__":

    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.push(5)

    stack.print()
    print(stack.pop())
    stack.print()